core/src/mast/merger/mod.rs

```diff
@@ -0,0 +1,429 @@
+use alloc::{collections::BTreeMap, vec::Vec};
+use core::ops::ControlFlow;
+
+use miden_crypto::hash::{blake::Blake3Digest, rpo::RpoDigest};
+
+use crate::mast::{DecoratorId, EqHash, MastForest, MastForestError, MastNode, MastNodeId};
+
+#[cfg(test)]
+mod tests;
+
+/// A type that allows merging [`MastForest`]s.
+///
+/// This functionality is exposed via [`MastForest::merge`]. See its documentation for more details.
+pub(crate) struct MastForestMerger {
+    mast_forest: MastForest,
+    // Internal indices needed for efficient duplicate checking and EqHash computation.
+    node_id_by_hash: BTreeMap<RpoDigest, Vec<(EqHash, MastNodeId)>>,
+    hash_by_node_id: BTreeMap<MastNodeId, EqHash>,
+    decorators_by_hash: BTreeMap<Blake3Digest<32>, DecoratorId>,
+}
+
+impl MastForestMerger {
+    /// Creates a new merger which creates a new internal, empty forest into which other
+    /// [`MastForest`]s are merged.
+    pub(crate) fn new() -> Self {
+        Self {
+            node_id_by_hash: BTreeMap::new(),
+            hash_by_node_id: BTreeMap::new(),
+            decorators_by_hash: BTreeMap::new(),
+            mast_forest: MastForest::new(),
+        }
+    }
+
+    /// Merges `other_forest` into the forest contained in self.
+    pub(crate) fn merge(
+        &mut self,
+        other_forest: &MastForest,
```

Comment:
We should be able to pass by value instead of reference here, since I don't expect we would ever need to merge a forest, and keep using the un-merged version? This would allow us to reduce the amount of copying (e.g. all of a `BasicBlock`'s operations can be moved instead of copied)


cc @bobbinth @bitwalker

Reply:
I think the places where we can truly avoid clones is
- `BasicBlockNode` operations
- `DecoratorId`s in nodes by mapping them in-place
- decorators of `MastForest`s

Some things got less nice though:
- The iterator needs to consume the forest and return the owned `MastNode` (instead of `&MastNode`) which requires replacing the node in the forest with another one (or unsafe code, but this seems like a can of worms here - dropping a `Vec<MastNode>` with potentially uninitialized memory would have to be handled explicitly). This is fine overall as we only visit each node once, but it's ugly code nonetheless.
- Uses of `core::mem::take` which are possible right now due to not accessing the taken parts later, but it adds some complexity.

This was implemented in https://github.com/0xPolygonMiden/miden-vm/pull/1534/commits/9394a3aa4a49e344053cb2aca0a80c92340e1131 so it can be viewed in isolation to judge whether it's better or worse.

Overall I'm okay with leaving it with that commit, as it probably does have some performance benefits.

Reply:
I left a couple of comments related to this. Overall, passing references is more flexible, I think, even though we'd give up some performance. One scenario i'm thinking about is merging MAST forests from a bunch of libraries - and here, I think, working with references should be easier.

Reply:
> Overall I'm okay with leaving it with that commit, as it probably does have some performance benefits.

Agreed, I think the benefits outweigh the negatives. Basic block nodes and decorators take up most of the memory in a `MastForest`, so cloning those is almost equivalent to cloning the whole thing - so I think it's best to avoid it.

> Overall, passing references is more flexible, I think, even though we'd give up some performance.

As mentioned above, we were effectively cloning the whole MAST forest internally, so when this is the case, I prefer giving the caller the choice between moving or cloning (i.e. only cloning if the object is needed after the call). As for flexibility, if all the caller has is a reference, then all it needs to do is `forest.merge(other_forest.clone())` - a `clone()` we were doing internally anyways.

Reply:
Due to Bobbin's comment and the refactoring (https://github.com/0xPolygonMiden/miden-vm/pull/1534#discussion_r1813086360) I've reverted my change for the merge to take ownership of the forest. This Multi Forest iterator implemented now holds a reference to the forest and doesn't return `MastNode`s itself, only indices. But even if we took ownership now of the forest, we couldn't move the nodes out of the merging forest because the iterator still holds an immutable reference to it. So I think without another refactoring we might be stuck on taking references now, although I haven't thought too much about it until now, because I just finished this refactor.

Reply:
Let's just leave it as-is then